### PR TITLE
Add support and validation for GPU in clusterConstraints

### PR DESCRIPTION
**Pull request by eshiroma**
/gcbrun

### CODE REVIEW

**Comment by eshiroma** [comment]
Also validate that `requests` is set on `clusterConstraints.resources`.

/gcbrun

File: marketplace/deployer_util/config_helper.py [hunk 1]
```diff
@@ -405,12 +414,33 @@ def minimum_node_count(self):
...
   """Accesses a single resource's requests."""
 
   def __init__(self, dictionary):
     self._cpu = dictionary.get('cpu', None)
     self._memory = dictionary.get('memory', None)
+    self._gpu = None
+
+    rawGpu = dictionary.get('gpu', None)
+    if rawGpu != None:
+      if not rawGpu.keys():
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper.py]: check the type to be a dict first

File: marketplace/deployer_util/config_helper.py [hunk 2]
```diff
@@ -405,12 +414,33 @@ def minimum_node_count(self):
...
+        raise InvalidSchema('GPU requests map must contain one or more entries')
+      for key in rawGpu.keys():
+        if key not in _GPU_PROVIDER_KEYS:
+          raise InvalidSchema('Unrecognized GPU provider %s', key)
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper.py]: s/Unrecognized/Unsupported

File: marketplace/deployer_util/config_helper.py [hunk 3]
```diff
@@ -354,8 +356,15 @@ def __init__(self, dictionary):
...
-    self._requests = _maybe_get_and_apply(
-        dictionary, 'requests', lambda v: SchemaResourceConstraintRequests(v))
+    self._requests = _must_get_and_apply(
+        dictionary, 'requests', lambda v: SchemaResourceConstraintRequests(v),
+        'Each item in clusterConstraints.resources must specify requests')
+
+    if self._requests.gpu:
+      if self._affinity:
+        raise InvalidSchema('Affinity unsupported for GPU resource constraints')
+      if self._replicas:
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper.py]: if gpu is specified, disallow cpu and memory as well?

File: marketplace/deployer_util/config_helper_test.py [hunk 4]
```diff
@@ -1038,15 +1038,26 @@ def test_resource_constraints(self):
...
                   type: REQUIRE_ONE_NODE_PER_REPLICA
             - replicas: 5
+              requests:
+                cpu: 50m
               affinity:
                 simpleNodeAffinity:
                   type: REQUIRE_MINIMUM_NODE_COUNT
                   minimumNodeCount: 4
+            - requests:
+                cpu: 200m
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper_test.py]: rm?

File: marketplace/deployer_util/config_helper.py [hunk 5]
```diff
@@ -354,8 +356,15 @@ def __init__(self, dictionary):
     self._replicas = dictionary.get('replicas', None)
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper.py]: I think replicas should be required, other than gpu requests

**Comment by eshiroma** [comment]
Partially addresses #483 (validates `requests` in `resources` entries)

/gcbrun

File: marketplace/deployer_util/config_helper_test.py (hunk 4 above)
Comment (huyhg) [reply marketplace/deployer_util/config_helper_test.py]: Remove `cpu` spec--we're don't have the ability to fit nodes that request both cpu and gpu.

Comment (eshiroma) [reply marketplace/deployer_util/config_helper_test.py]: Done, as a part of disallowing cpu/memory constraints if GPU requests are specified.

File: marketplace/deployer_util/config_helper.py (hunk 2 above)
Comment (eshiroma) [reply marketplace/deployer_util/config_helper.py]: Done.

**Pull request merged**